src/AutoRest.CSharp/Common/Output/Models/Types/HelperTypeProviders/System/ClientPipelineExtensionsProvider.cs

```diff
@@ -0,0 +1,197 @@
+            {
+                new InvokeInstanceMethodStatement(_pipeline, nameof(Pipeline<PipelineMessage>.Send), new[]{ _message }, false),
+                GetProcessMessageBody()
+            });
+        }
+
+        private MethodBodyStatement GetProcessMessageBody()
+        {
+            return new MethodBodyStatement[]
+            {
+                EmptyLine,
+                new IfStatement(Equal(_messageResponse, Null))
+                {
+                    Throw(New.InvalidOperationException(Literal("Failed to receive Result.")))
+                },
+                EmptyLine,
+                new IfStatement(Not(new BoolExpression(new MemberExpression(_messageResponse, "IsError"))).Or(Equal(new MemberExpression(_requestContext, "ErrorBehavior"), FrameworkEnumValue(ErrorBehavior.NoThrow))))
+                {
+                    Return(_messageResponse)
+                },
+                EmptyLine,
+                Throw(new NewInstanceExpression(typeof(MessageFailedException), new[] { _messageResponse }))
+            };
+        }
+
+        private MethodSignature GetProcessMessageSignature(bool isAsync, out Parameter cancellationTokenParam)
+        {
+            cancellationTokenParam = new Parameter("cancellationToken", null, typeof(CancellationToken), Constant.Default(typeof(CancellationToken)), ValidationType.None, null);
+            var modifiers = MethodSignatureModifiers.Public | MethodSignatureModifiers.Static | MethodSignatureModifiers.Extension;
+            if (isAsync)
+            {
+                modifiers |= MethodSignatureModifiers.Async;
+            }
+            return new MethodSignature(
+                isAsync ? "ProcessMessageAsync" : "ProcessMessage",
+                null,
+                null,
+                modifiers,
+                isAsync ? typeof(ValueTask<PipelineResponse>) : typeof(PipelineResponse),
+                null,
+                new[] { _pipelineParam, _messageParam, _requestContextParam, cancellationTokenParam });
+        }
+
+        private Method BuildProcessMessageAsync()
+        {
+            MethodSignature signature = GetProcessMessageSignature(true, out var cancellationTokenParam);
+            var cancellationToken = new ParameterReference(cancellationTokenParam);
+            return new Method(signature, new MethodBodyStatement[]
+            {
+                new InvokeInstanceMethodStatement(_pipeline, nameof(Pipeline<PipelineMessage>.SendAsync), new[]{ _message }, true),
+                GetProcessMessageBody()
+            });
+        }
+
+        internal PipelineResponseExpression ProcessMessage(List<ValueExpression> arguments, bool isAsync)
+        {
+            return new(new InvokeStaticMethodExpression(Type, isAsync ? _processMessageAsync : _processMessage, arguments, CallAsExtension: true, CallAsAsync: isAsync));
+        }
+
+        internal ResultExpression ProcessHeadAsBoolMessage(List<ValueExpression> arguments, bool async)
```

Comment:
last argument is called `async` in this method but `isAsync` in `ProcessMessage`. Let's make this consistent.
Also, `arguments` can be `IReadOnlyList<ValueExpression>`